Automattic/jetpack
# eslint-changed: Change to esm and use jest for tests

#### Changes proposed in this Pull Request:

The ESM switch was very straightforward, all that was needed was to swap
`require` for `import` and replace some uses of `__dirname` in the tests.

The switch to `jest` was a little more complicated:

* The first part was simple enough, once `--experimental-vm-modules` was
  added then it worked without even disabling code transforms.
* But jest's resolver choked on a fancy feature in the new version of
  chalk, so a workaround had to be added for that. Sigh.
* And then getting coverage working required a restructure of all the
  code, because jest won't record coverage for code in child processes.

And then when I was getting all the tests to work with the restructured
code, I noticed that "Filters by filename" was testing for incorrect
behavior (it wasn't actually filtering). So I had to fix that too.

#### Jetpack product discussion


The switching to jest part is relevant to p4TIVU-a8r-p2#comment-11500

The esm part is just something I've wanted to do for a while now, and it helped me figure out how to get jest to run in an esm project.

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Is CI happy?
* Try out eslint-changed in other ways, does it still work?

P.S. GitHub's UI is probably not going to be very helpful with the move of most of the code from `bin/eslint-changed.js` to `src/cli.js`. To better see the changes there you might do like `git diff -w --minimal origin/master:projects/js-packages/eslint-changed/bin/eslint-changed.js projects/js-packages/eslint-changed/src/cli.js`
